Create runtime staging clone to manually kick off full test runs

steveisok opened the pull request:
This change allows devs to manually kick off full test runs on the configurations that only execute smoke tests per PR.

`/azp run runtime-staging-manual` will do the trick

steveisok committed:
Create runtime staging clone to manually kick off full test runs

[skip ci]

safern commented on eng/pipelines/manual-runtime-staging.yml:
```diff
@@ -0,0 +1,435 @@
+trigger: none
```
A small suggestion, rather than using a custom yml file, could we use the existing file? Is this essentially identical to `runtime-staging`? Or will this set `isFullMatrix=true`?

safern commented:
/azp run manual-runtime-staging

Commit:
Remove separate pipeline

Commit:
Tweak variables to support full matrix on manual runs

steveisok commented:
/azp run manual-runtime-staging

safern commented on eng/pipelines/common/variables.yml:
```diff
@@ -12,9 +12,9 @@ variables:
   value: ${{ and(ne(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest')) }}
 - name: isFullMatrix
-  value: ${{ and(eq(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest')) }}
+  value: ${{ or(in(variables['Build.DefinitionName'], 'manual-runtime-staging', 'manual-runtime-staging'), and(eq(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest'))) }}
```
```suggestion
  value: ${{ or(in(variables['Build.DefinitionName'], 'manual-runtime-staging', 'manual-runtime'), and(eq(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest'))) }}
```

steveisok committed:
Update eng/pipelines/common/variables.yml

Co-authored-by: Santiago Fernandez Madero <safern@microsoft.com>

safern commented on eng/pipelines/common/variables.yml:
```diff
@@ -12,9 +12,9 @@ variables:
   value: ${{ and(ne(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest')) }}
 - name: isFullMatrix
-  value: ${{ and(eq(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest')) }}
+  value: ${{ or(in(variables['Build.DefinitionName'], 'manual-runtime-staging', 'manual-runtime'), and(eq(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest'))) }}
```
I would be careful of adding a pipeline that runs with `isfullMatrix=true` for `runtime.yml` as that burns a lot of resources and runs a ton of stuff (it is equivalent to our rolling builds which we just reduced to 2 a day), so maybe we can be more granular there?

steveisok replied:
yeah, I think it can go in another variable.

Commit:
Add a separate isSmokeTestBuild variable and trickle it into runtime-staging

Commit:
Merge remote-tracking branch 'upstream/stpfiste/manual-runtime-staging' into stpfiste/manual-runtime-staging

steveisok commented:
/azp run manual-runtime-staging

Commit:
Merge remote-tracking branch 'upstream/main' into stpfiste/manual-runtime-staging

steveisok commented:
/azp run manual-runtime-staging

Commit:
Flip conditions

steveisok commented:
/azp run manual-runtime-staging

radical commented on eng/pipelines/common/variables.yml:
```diff
@@ -15,6 +15,10 @@ variables:
   value: ${{ and(eq(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest')) }}
 - name: isNotFullMatrix
   value: ${{ and(eq(variables['System.TeamProject'], 'public'), eq(variables['Build.Reason'], 'PullRequest')) }}
+- name: isSmokeTestBuild
+  value: ${{ and(not(in(variables['Build.DefinitionName'], 'manual-runtime-staging', 'manual-runtime')), eq(variables['System.TeamProject'], 'public'), eq(variables['Build.Reason'], 'PullRequest')) }}
```
```
isSmokeTestbuild    = (!manual && PR)
isNotSmokeTestbuild = (manual || !PR)
```
Should we name these two to be indicative of when they would be true, like `isNotManualAndIsPR`, and `isManualOrIsPR`? What to do based on this can then be determined where they are used, for example, for `RunSmokeTestOnly=$(isNotManualPR)`.

radical replied:
It would also make it easier to read conditions:

```yml
...
           eq(variables['isNotSmokeTestBuild'], true),
           eq(variables['isFullMatrix'], true))
...
```

.. vs ..

```yml
...
           eq(variables['isNotManualAndIsPr'], true),
           eq(variables['isFullMatrix'], true))
...
```

steveisok replied:
I'm open to ideas :-)

Commit:
Naming feedback

steveisok commented:
/azp run manual-runtime-staging

akoeplinger commented on eng/pipelines/common/variables.yml:
```diff
@@ -15,6 +15,10 @@ variables:
+- name: isNotManualAndIsPR
+  value: ${{ and(not(in(variables['Build.DefinitionName'], 'manual-runtime-staging', 'manual-runtime')), eq(variables['System.TeamProject'], 'public'), eq(variables['Build.Reason'], 'PullRequest')) }}
+- name: isManualOrIsNotPR
+  value: ${{ or(in(variables['Build.DefinitionName'], 'manual-runtime-staging', 'manual-runtime'), and(eq(variables['System.TeamProject'], 'public'), ne(variables['Build.Reason'], 'PullRequest'))) }}
```
nit: can we rename the pipeline to `runtime-staging-manual`? that would sort alphabetically next to the existing runtime-staging which I think makes more sense.

akoeplinger commented on eng/pipelines/common/variables.yml:
```diff
@@ -15,6 +15,10 @@ variables:
   value: ${{ and(eq(variables['System.TeamProject'], 'public'), eq(variables['Build.Reason'], 'PullRequest')) }}
+- name: isNotManualAndIsPR
+  value: ${{ and(not(in(variables['Build.DefinitionName'], 'manual-runtime-staging', 'manual-runtime')), eq(variables['System.TeamProject'], 'public'), eq(variables['Build.Reason'], 'PullRequest')) }}
+- name: isManualOrIsNotPR
```
I'd call the variable `isReducedFullMatrix` since we're essentially running a bit less than the normal fullmatrix run, I don't think we need to tie this into PR vs. rolling vs. manual.

akoeplinger replied:
alternatively, I wonder if we should instead just set isFullMatrix for the manual pipeline and only update the few places where we really really don't want to run on PRs even in the manual pipeline.

steveisok replied:
> alternatively, I wonder if we should instead just set `isFullMatrix` for the manual pipeline and only update the few places where we really really don't want to run on PRs even in the manual pipeline.

For runtime-staging that might be ok, but I think it would be harder to work around in runtime.  That's why I thought Santi's suggestion for another variable was a good one.

steveisok replied:
> I'd call the variable `isReducedFullMatrix` since we're essentially running a bit less than the normal fullmatrix run, I don't think we need to tie this into PR vs. rolling vs. manual.

I thought Ankit's suggestion was easier to work with when modifying the yml.  Honestly, I don't care what we name it.

@radical @safern, thoughts?

akoeplinger replied:
When I quickly looked it didn't seem used in a lot of relevant places but ok :)

That actually raises another question, or maybe I'm misunderstanding the conditions: we're running _all_ of the legs in the manual pipeline, even those that already run as part of the normal pipeline right? that seems suboptimal and a waste of resources.

steveisok replied:
We'll only run a subset of the legs.  Basically, all the ones that are configured to run smoke tests by default.

safern replied:
My biggest concern with using isFullMatrix is for runtime.yml because that runs as a rolling build which has a LOOT of tests and also a lot of those would already be covered by the default run we get on PRs, so that's why I suggested a new variable.

steveisok replied:
@akoeplinger @radical @safern Are the `isNotManualAndIsPR` and `isManualOrIsNotPR` variables ok or do we want a different name?  My vote is to go with them, but I'd like your opinions before moving forward.

radical replied:
I would prefer sticking with them, but would be fine with whatever @safern, and @akoeplinger agree upon.

akoeplinger replied:
I'm fine with sticking with them too.

safern replied:
Fine with sticking with those names as well.

steveisok commented:
/azp runtime-staging-manual

steveisok commented:
/azp run runtime-staging-manual

Commit:
Flipped name of manual pipeline

steveisok commented:
/azp run runtime-staging-manual

Commit:
Prevent anything being scheduled on the manual pipelines

steveisok commented:
/azp run runtime-staging-manual

akoeplinger commented on eng/pipelines/runtime-staging.yml:
```diff
@@ -22,13 +22,14 @@ trigger:
-      include:
-      - main
-    always: false # run only if there were changes since the last successful scheduled run.
+- ${{ if not(in(variables['Build.DefinitionName'], 'runtime-staging-manual', 'runtime-manual') }}:
```
nit: `runtime-manual` doesn't apply here since this is runtime-staging.yml

steveisok commented:
/azp run runtime-staging-manual

steveisok commented:
@safern @akoeplinger conditioning the schedule based on pipeline name doesn't seem to work.  Any suggestions?  I think this might be where we actually need a separate yml.

```
- ${{ if not(eq(variables['Build.DefinitionName'], 'runtime-staging-manual')) }}:
  schedules:
    - cron: "0 7,19 * * *" # run at 7:00 and 19:00 (UTC) which is 23:00 and 11:00 (PST).
      displayName: Runtime-staging default schedule
      branches:
        include:
        - main
      always: false # run only if there were changes since the last successful scheduled run.
```

akoeplinger committed:
Use template for runtime-staging so we can extract out the schedule and trigger parts

akoeplinger commented:
/azp run runtime-staging-manual

safern commented on eng/pipelines/runtime-staging-manual.yml:
```diff
@@ -0,0 +1,4 @@
+trigger: none
+
+extends:
+    template: runtime-staging-template.yml
```
now that we have a template, maybe we can use parameters to turn on/off legs rather than checking based on pipeline name?

radical replied:
Like platform!

akoeplinger replied:
that works for runtime-staging but we still have runtime.yml, so I'd need to move that to template as well which I'm not sure is desired.

safern replied:
That one also has a schedule trigger, so it will limit us as well and we will need the template anyway?

steveisok replied:
I think this is a good first step.  I think we can get more custom in a follow up PR.

Commit:
Exclude DllImportantGenerator Unit tests from mobile

Commit:
Make sure a few more configurations run manually
